cachito/workers/nexus.py

```diff
@@ -149,3 +151,146 @@ def get_ca_cert():
+            format=component_format, group=group, name=name, repository=repository, version=version
+        )
+        if len(components) > 1:
+            raise CachitoError(
+                "The component search in Nexus unexpectedly returned more than result"
+            )
+        if components:
+            return components[0]
+
+        attempts += 1
+
+    return None
+
+
+def search_components(**query_params):
+    """
+    Search for components using the Nexus REST API.
+
+    :param query_params: the query parameters to filter
+    :return: the list of components returned by the search
+    :rtype: list<dict>
+    :raise CachitoError: if the search fails
+    """
+    # Import this here to avoid a circular import
+    from cachito.workers.requests import requests_session
+
+    config = get_worker_config()
+    auth = requests.auth.HTTPBasicAuth(config.cachito_nexus_username, config.cachito_nexus_password)
+    url = f"{config.cachito_nexus_url.rstrip('/')}/service/rest/v1/search"
+    # Create a copy so that the original query parameters are unaltered later on
+    params = copy.deepcoams = copy.deepcopy(query_params)
+
+    log.debug(
+        "Searching Nexus for components using the following query parameters: %r", query_params
+    )
+    items = []
+    while True:
+        try:
+            rv = requests_session.get(
+                url, auth=auth, params=params, timeout=config.cachito_nexus_timeout
+            )
+        except requests.RequestException:
+            msg = "Could not connect to the Nexus instance to search for components"
+            log.exception(msg)
+            raise CachitoError(msg)
+
+        if not rv.ok:
+            log.error(
+                "Failed to search for components (%r) in Nexus with the status code %d and the "
+                "text: %s",
+                query_params,
+                rv.status_code,
+                rv.text,
+            )
+            raise CachitoError("Failed to search for components in Nexus")
+
+        rv_json = rv.json()
+        items.extend(rv_json["items"])
+
+        # Handle pagination
+        if rv_json["continuationToken"]:
+            log.debug("Getting the next page of Nexus component search results")
+            params["continuationToken"] = rv_json["continuationToken"]
+        else:
+            break
+
+    return items
+
+
+def upload_artifact(repo_name, repo_type, artifact_path):
+    """
+    Upload an artifact to the Nexus hosted repository.
+
+    :param str repo_name: the name of the Nexus hosted repository
+    :param str repo_type: the type of the Nexus hosted repository (e.g. ``npm``)
+    :param str artifact_path: the path to the artifact to upload
+    :raise CachitoError: if the upload fails
+    """
+    # Import this here to avoid a circular import
+    from cachito.workers.requests import requests_session
+
+    with open(artifact_path, "rb") as artifact:
+        file_payload = {f"{repo_type}.asset": artifact.read()}
```

Comment:
We may want to consider using a stream instead of loading the whole binary into memory. This can be done later on.